### PR TITLE
76 search

### PR DESCRIPTION
**Pull request by danylo-dudok**

### CODE REVIEW

File: TransIT.API/Controllers/DataController.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    [ApiController]
+    [EnableCors("CorsPolicy")]
+    [Produces("application/json")]
+    [Route("api/v1/[controller]/[action]")]
+    public abstract class DataController<TEntity, TEntityDTO> : Controller
+        where TEntity : class, IEntity, new()
+        where TEntityDTO : class, new()
+    {
+        private readonly ICrudService<TEntity> _dataService;
+        protected readonly IMapper Mapper;
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/DataController.cs]: mapper
or use Mapper {get; set;}

File: TransIT.API/Controllers/DataController.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            if (ModelState.IsValid)
+            {
+                var res = await _dataService.GetRangeAsync(offset, amount);
+                if (res != null) return Json(res);
+            }
+            return BadRequest();
+        }
+
+        [HttpGet]
+        public virtual async Task<IActionResult> Get([FromQuery] int id)
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/DataController.cs]: HttpGet("{id}")

File: TransIT.API/Controllers/DataController.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        protected readonly IMapper Mapper;
+        
+        public DataController(IMapper mapper, ICrudService<TEntity> dataService)
+        {
+            Mapper = mapper;
+            _dataService = dataService;
+        }
+
+        [HttpGet]
+        public virtual async Task<IActionResult> Get([FromQuery] uint offset, uint amount)
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/DataController.cs:29]: Change route

Comment (CrazyRoka) [reply TransIT.API/Controllers/DataController.cs:29]: Make it more restfull

File: TransIT.API/Controllers/DataController.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        {
+            if (ModelState.IsValid)
+            {
+                var res = await _dataService.GetRangeAsync(offset, amount);
+                if (res != null) return Json(res);
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/DataController.cs]: Convert to DTO

File: TransIT.API/Controllers/DataController.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            return BadRequest();
+        }
+
+        [HttpGet]
+        public virtual async Task<IActionResult> Get([FromQuery] int id)
+        {
+            if (ModelState.IsValid)
+            {
+                var res = await _dataService.GetAsync(id);
+                if (res != null) return Json(res);
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/DataController.cs]: Convert to dto

File: TransIT.API/Controllers/DataController.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            return BadRequest();
+        }
+
+        [HttpGet]
+        public virtual async Task<IActionResult> Get([FromQuery] string search)
+        {
+            if (ModelState.IsValid)
+            {
+                var res = await _dataService.SearchAsync(search);
+                if (res != null) return Json(res);
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/DataController.cs]: Convert to dto
search.Trim()

File: TransIT.API/Controllers/DataController.cs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        public virtual async Task<IActionResult> Create([FromBody] TEntityDTO obj)
+        {
+            if (ModelState.IsValid)
+            {
+                var entity = await _dataService.CreateAsync(
+                    Mapper.Map<TEntity>(obj));
+                var res = Mapper.Map<TEntityDTO>(entity);
+                
+                if (res != null)
+                    return Created($"{Request.Path.Value}/{entity.Id.ToString()}", res);
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/DataController.cs]: Use CreatedAtRoute

File: TransIT.API/Controllers/DataController.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            }
+            return BadRequest();
+        }
+
+        [HttpPut]
+        public virtual async Task<IActionResult> Update([FromBody] TEntityDTO obj)
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/DataController.cs]: {id}

File: TransIT.API/Controllers/DataController.cs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+                var res = Mapper.Map<TEntityDTO>(entity);
+
+                if (res != null)
+                    return NoContent();
+            }
+            return BadRequest();
+        }
+
+        [HttpDelete]
+        public virtual async Task<IActionResult> Delete([FromQuery] int id)
```
Comment (CrazyRoka) [inline TransIT.API/Controllers/DataController.cs]: {id}

File: TransIT.BLL/Services/ImplementedServices/SupplierService.cs [hunk 10]
```diff
@@ -23,5 +26,21 @@ public class SupplierService : CrudService<Supplier>, ISupplierService
...
+            {
+                return _unitOfWork.SupplierRepository.GetAllAsync(a =>
+                    a.Name.ToUpperInvariant().Contains(search)
+                    || search.Contains(a.Name.ToUpperInvariant()));
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, nameof(SearchAsync));
+                return null;
+            }
```
Comment (CrazyRoka) [inline TransIT.BLL/Services/ImplementedServices/SupplierService.cs]: Don't Repeat Yourself

**Review by CrazyRoka — DISMISSED** [review]

**Review by maksym-yaremko — APPROVED** [review]

**Review by markiyanshvets — APPROVED** [review]

**Pull request merged**